lending/sources/lending.move

```diff
@@ -0,0 +1,375 @@
+        if (!vector::contains(&vault.collaterals, &market_obj)) {
+            vector::push_back(&mut vault.collaterals, market_obj);
+        };
+
+        ctoken
+    }
+
+    public fun redeem(
+        redeemer: address,
+        market_obj: Object<Market>,
+        ctoken_fa: FungibleAsset
+    ): FungibleAsset acquires Market, Vault {
+        assert!(ctoken_metadata(market_obj) == fungible_asset::asset_metadata(&ctoken_fa), ERR_MARKET_MISMATCH);
+
+        // burn ctoken
+        let ctoken_amount = fungible_asset::amount(&ctoken_fa);
+        let underlying_amount = ctoken_to_underlying(market_obj, ctoken_amount);
+        let market = borrow_global<Market>(object::object_address(&market_obj));
+        fungible_asset::burn(&market.ctoken_burn_ref, ctoken_fa);
+
+        // update market fungible store
+        let market_signer = object::generate_signer_for_extending(&market.extend_ref);
+        let underlying = fungible_asset::withdraw(&market_signer, market_obj, underlying_amount);
+
+        let (_, shortfall) = account_liquidity(redeemer);
+        assert!(shortfall == 0, ERR_SHORTFALL);
+
+        underlying
+    }
+
+    public fun borrow(
+        borrower: address,
+        market_obj: Object<Market>,
+        amount: u64
+    ): FungibleAsset acquires Market, Vault {
+        // update market fungible store and total_borrow
+        let market = borrow_global_mut<Market>(object::object_address(&market_obj));
+        let total_borrow = &mut market.total_borrow;
+        *total_borrow = *total_borrow + amount;
+        let market_signer = object::generate_signer_for_extending(&market.extend_ref);
+        let fa = fungible_asset::withdraw(&market_signer, market_obj, amount);
+
+        // update user vault
+        let vault = borrow_global_mut<Vault>(borrower);
+        if (!simple_map::contains_key(&vault.debts, &market_obj)) {
+            simple_map::add(&mut vault.debts, market_obj, amount);
+        } else {
+            let debt = simple_map::borrow_mut(&mut vault.debts, &market_obj);
+            *debt = *debt + amount;
+        };
+
+        let (_, shortfall) = account_liquidity(borrower);
+        assert!(shortfall == 0, ERR_SHORTFALL);
+
+        fa
+    }
+
+    public fun repay(repayer: address, market_obj: Object<Market>, fa: FungibleAsset) acquires Vault, Market {
+        assert!(fungible_asset::asset_metadata(&fa) == fungible_asset::store_metadata(market_obj), ERR_MARKET_MISMATCH);
+
+        let amount = fungible_asset::amount(&fa);
+
+        // update market fungible store
+        fungible_asset::deposit(market_obj, fa);
+        let market = borrow_global_mut<Market>(object::object_address(&market_obj));
+        let total_borrow = &mut market.total_borrow;
+        *total_borrow = *total_borrow - amount;
+
+        // update user vault
+        let vault = borrow_global_mut<Vault>(repayer);
+        let debt = simple_map::borrow_mut(&mut vault.debts, &market_obj);
+        *debt = *debt - amount;
```

Comment:
I think it's better to take `&mut FungibleAsset` and take min(fa, debt). Remove debt if becomes 0.